### PR TITLE
feat: Always add LIMIT 1 to `findOne` queries

### PR DESCRIPTION
**Pull request by rharriso**
### Pull Request Checklist

_Please make sure to review and check all of these items:_

- [x] Have you added new tests to prevent regressions? **No, tests remove**
- [X] Does `yarn test` or `yarn test-DIALECT` pass with this change (including linting)?
- [x] Is a documentation update included (if this change modifies existing APIs, or introduces new ones)? No this removes an undocumented behavior
- [x] Did you update the typescript typings accordingly (if applicable)?
- [x] Does the description below contain a link to an existing issue (Closes #[issue]) or a description of the issue you are solving?
- [x] Did you follow the commit message conventions explained in [CONTRIBUTING.md](https://github.com/sequelize/sequelize/blob/main/CONTRIBUTING.md)?

### Description Of Change

Remove conditional logic for adding `LIMIT 1` to `findOne` queries. Instead, always add it.

* Foreign keys should be used to ensure data integrity, and should not impact query performance.
* It's unclear how omitting `LIMIT 1` benefits any query that should return a single result.
* Models are imperfect representations of schema, making unnecessarily assumptions about their correctness can lead to severe issues.
  * see #14548

### CODE REVIEW

**Comment by kpeluso** [comment]
+1

what is being corrected here is an "anti-feature"

**Comment by ephys** [comment]
I agree, not adding `LIMIT 1` if we're comparing the primary key sounds weird to me. I'd expect `Model.findOne()` to be equivalent to `(await Model.findAll({ limit: 1 }))[0]`.

Strangely enough, some mssql tests are failing with this change

**Comment by rharriso** [comment]
@ephys , sorry to say I have been able to identify a commit where these tests started failing. I'm wondering if that has come up in other PRs?

To Reiterate, the MSSQL tests fail locally for me on `main`

**Comment by ephys** [comment]
are you using `start-mssql` to start the server?

**Comment by rharriso** [comment]
@ephys, I'm trying to make sure I'm properly running the test locally. 
I'll get back to you once I'm certain that I'm getting the same resuts as CI.

Thanks for your patience.

File: src/model.js [hunk 1]
```diff
@@ -1875,18 +1875,8 @@ ${associationOwner._getAssociationDebugList()}`);
     }
 
     options = Utils.cloneDeep(options);
-
-    if (options.limit === undefined) {
```
Comment (WikiRik) [inline src/model.js:1879]: I'm not sure why but keeping this if statement fixes the failing tests.
```js
if (options.limit === undefined) {
  options.limit = 1;
}
```
seems to work just fine

Comment (rharriso) [reply src/model.js:1879]: One of the failing tests is below.
For some reason the MSSQL query is including the `User.id` column in the query even after it is deleted on line 666 🤘.

Logging the sql string in `abstract/query-interface.js#select`:

```sql
SELECT [id], [title], [createdAt], [updatedAt], [user_name] FROM [Tasks] AS [Task] WHERE [Task].[title] = N'some task' ORDER BY [Task].[id] OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;
SELECT [username], [createdAt], [updatedAt] FROM [Users] AS [User] WHERE [User].[username] = N'bob' ORDER BY [User].[id] OFFSET 0 ROWS FETCH NEXT 1 ROWS ONLY;
```

Notice `[User].[id]` in the order statement.

https://github.com/sequelize/sequelize/blob/428a85b1376b09373a362d42abedfae0790b8293/test/integration/associations/belongs-to.test.js#L662-L682

Comment (rharriso) [reply src/model.js:1879]: Dug a bit deeper. And I think I found the source of the issue. 

`src/dialects/mssql/query-generator.js#addLimitAndOffeset`.

`primaryKeyField` is used to order the subqueries. You'll notice in my comment above that the primary key would have been deleted.

See the following console output.

```
console.log('shouldn't be', Object.keys(model.rawAttributes).join(' 🤝 '));
shouldn't be username 🤝 createdAt 🤝 updatedAt
console.log('still here','${this.quoteTable(options.tableAs || model.name)}.${this.quoteIdentifier(model.primaryKeyField)}'
still here [User].[id]
'😭 ${model.primaryKeyField}'
😭 id
```

Not quite sure what the resolution is here. If the PK has been deleted.....

**Comment by rharriso** [comment]
I've added the conditional pointed out by @WikiRik, and opened up issue #14618

**Comment by ephys** [comment]
I found the reason why LIMIT was not included https://github.com/sequelize/sequelize/pull/9248. I'm documenting it here for future PRs but I disagree with the premise of that PR and I agree LIMIT should be specified.

**Review by ephys — APPROVED** [review]

**Review by WikiRik — APPROVED** [review]

**Comment by WikiRik** [comment]
I'll run the tests again after GitHub solves their incident; https://www.githubstatus.com/incidents/6r4sms3hcj3x

**Comment by vanthome** [comment]
Hello, I have my doubts that this is correct. Are we sure that `findOne` queries should really only ever return one row? We are using `findOne` in conjunction with includes and then, because of outer joins happening, we do get and expect multiple rows. Please correct me if I'm wrong. Thx

**Comment by ephys** [comment]
Hi :)

In `find{One,All}`, `limit` doesn't necessarily mean only one row in the database is going to be fetched. We want `limit` to only apply to the level it's on, in the redesign. So if you wrote the following:

```ts
User.findOne({
  include: [User.associations.projects],
});
```

It would return a single user, but include all of the user's projects

We know that currently `limit` behaves in very inconsistent ways, but it's something [I'm working on addressing](https://github.com/sequelize/website/pull/298/files#diff-d8aa946ffed5a9f3e0abc36699e64393e20f1ad6ca5b0997aa165325f3fa651cR324)

**Pull request merged**